crypto/threads_pthread.c

```diff
@@ -42,12 +44,480 @@
+    uint64_t users;
+};
+
+/*
+ * This is the per thread tracking data
+ * that is assigned to each thread participating
+ * in an rcu qp
+ *
+ * qp points to the qp that it last acquired
+ *
+ * count is the threads current recursion count
+ */
+struct rcu_thr_data {
+    struct rcu_qp *qp;
+    int count;
+};
+
+/*
+ * This is the internal version of a CRYPTO_RCU_LOCK
+ * it is cast from CRYPTO_RCU_LOCK
+ */
+struct rcu_lock_internal {
+    CRYPTO_RCU_LOCK lock; /* Must be first */
+    struct rcu_cb_item *cb_items;
+    uint64_t id_ctr;
+    struct rcu_qp *qp_group;
+    size_t group_count;
+    uint64_t reader_idx;
+    uint32_t next_to_retire;
+    uint64_t current_alloc_idx;
+    uint32_t writers_alloced;
+    pthread_mutex_t write_lock;
+    pthread_mutex_t alloc_lock;
+    pthread_cond_t alloc_signal;
+    pthread_mutex_t prior_lock;
+    pthread_cond_t prior_signal;
+};
+
+/*
+ * Called on thread exit to free the pthread key
+ * associated with this thread, if any
+ */
+static void free_rcu_thr_data(void *ptr)
+{
+    struct rcu_thr_data *data = 
+                        (struct rcu_thr_data *)CRYPTO_THREAD_get_local(&rcu_thr_key);
+
+    if (data)
+        CRYPTO_free(data, __FILE__, __LINE__);
+    CRYPTO_THREAD_set_local(&rcu_thr_key, NULL);
+}
+
+static void ossl_rcu_init(void)
+{
+    CRYPTO_THREAD_init_local(&rcu_thr_key, NULL);
+}
+
+/* Read side acquisition of the current qp */
+static struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK *lock)
+{
+    uint64_t qp_idx;
+    struct rcu_lock_internal *ilock =
+        (struct rcu_lock_internal *)lock;
+
+    /* get the current qp index */
+    for(;;) {
+        qp_idx = (uint64_t)__ATOMIC_LOAD_N(&ilock->reader_idx, __ATOMIC_SEQ_CST);
+        __ATOMIC_ADD_FETCH(&ilock->qp_group[qp_idx].users, VAL_READER,
+                           __ATOMIC_SEQ_CST);
+        /* if the idx hasn't changed, we're good, else try again */
+        if (qp_idx == (uint64_t)__ATOMIC_LOAD_N(&ilock->reader_idx, __ATOMIC_SEQ_CST))
+            break;
+
+        __ATOMIC_SUB_FETCH(&ilock->qp_group[qp_idx].users, VAL_READER,
+                           __ATOMIC_SEQ_CST);
+    }
+
+    return (struct rcu_qp *)&ilock->qp_group[qp_idx];
+}
+
+/* Public READ lock api */
+void ossl_rcu_read_lock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_thr_data *data;
+
+    /*
+     * we're going to access current_qp here so ask the
+     * processor to fetch it
+     */
+    data = CRYPTO_THREAD_get_local(&rcu_thr_key);
+
+    if (data == NULL) {
+        for (;;) {
+            data = OPENSSL_zalloc(sizeof(struct rcu_thr_data));
+            if (data != NULL)
+                break;
+        }
+        CRYPTO_THREAD_set_local(&rcu_thr_key, data);
+        ossl_init_thread_start(NULL, NULL, free_rcu_thr_data);
+    }
+
+    if (data->qp == NULL)
+        data->qp = get_hold_current_qp(lock);
+
+    /* inc our local count */
+    data->count++;
+
+    /* check for underflow condition */
+    assert(data->count >= 0);
```

Comment:
Isn't this a check for overflow?

Reply:
underflow, but yes.   when we call this functon, data->count should never be less than zero, so if that assert fails, someone previously has decremented data->count when they shouldn't have

Reply:
Err...no my point is you really are checking overflow not underflow (so the comment is wrong). You just incremented data->count, so the only way that assert will fail is if data->count has overflowed and is now negative. Assuming that it was >=0 to start with. If you wanted to check that we hadn't previously underflowed then I would expect the assert to be before the increment of data->count. But that doesn't make much sense to me